### ISSUE TITLE
retrieving a list of "abstracted" entities

### ISSUE THREAD

**Issue by jmfeurprier** [mapado/rest-client-sdk#55]
Hi there,

We have an entity hierarchy as such (simplified) :

- abstract Document { id, title, author }
- Article extends Document { content }
- Recipe extends Document { ingredients, calories }

We have an API end-point for documents which allows to make queries based on shared properties (id, title, author) from the Document base class. Here is a sample response for GET /documents :

```
{
  "@context": "/contexts/Document",
  "@id": "/documents",
  "@type": "hydra:Collection",
  "hydra:member": [
    {
      "@id": "/articles/320d346d-0c8f-11e7-8f02-0242ac150002",
      "@type": "Article",
      "id": "320d346d-0c8f-11e7-8f02-0242ac150002",
      "title": "I am an article",
      "author": "Me",
      "content": "article content"
    },
    {
      "@id": "/recipes/320d45f8-0c8f-11e7-8f02-0242ac150002",
      "@type": "Recipe",
      "id": "320d45f8-0c8f-11e7-8f02-0242ac150002",
      "title": "I am a recipe",
      "author": "You",
      "ingredients": "food and stuff",
      "calories": 5000
    }
  ],
  "hydra:totalItems": 2,
  "hydra:search": {
    // ...
  }
}
```

And then, when we try to call this end-point with the SDK client, the model hydrator explodes:

```
Cannot instantiate abstract class AppBundle\Entity\Document

Stack Trace
    in vendor/mapado/rest-client-sdk/src/Model/Serializer.php at line 84  -
                $identifierAttribute = $classMetadata->getIdentifierAttribute();
                $identifierAttrKey = $identifierAttribute ? $identifierAttribute->getSerializedKey() : null;
                $instance = new $className();
                foreach ($data as $key => $value) {
                    $attribute = $classMetadata->getAttribute($key);
                }
```

Of course, the Document is not to be instantiated as it is abstract. The model hydrator receives the "entityName" from an entity repository (DocumentRepository in our case), and then asks to hydrate the list of entities with this entityName, but in fact it should use the "@type" of the collection items to instantiate each entity, and maybe just use the entity name to validate the instantiated entities?

What do you think, is there a way to make it work?

**Comment by jmfeurprier** [comment]
I just found this comment in the serializer classs, line 79 :

`// classname may be detected for hydra api with @type key`

I guess this explains our issue :)

**Comment by jdeniau** [comment]
I do not have a great idea about this:

The fact is that the entity is really coupled with the url you use to make the call for now. 

A solution could be to allow a callback function to be called to override the default conversion functionality, but it may be complex and lead to bugs and be very coupled with the library, so more BC. 

Another solution is to implement a "kind of" inheritance system, like doctrine does but it may be really complex and add a lot of configuration. 

If you have a better idea, I will gladly take it as you are more close to the code.

**Comment by jmfeurprier** [comment]
I just pushed PR #56 , which is a proposal for the handling of abstract entities, let me know all the adjustments you'd like.

I added unit tests but I never used atoum before, my tests may not be 100% on par with your habits, just let me know.

**Comment by jdeniau** [comment]
I reviewed you PR, but I think of a nice "quickfix" you can do right now in your code before we fixed it in the API:

You can just override the `findBy` method like defined in [the README](https://github.com/mapado/rest-client-sdk/blob/bc9af2186daca9978ecfa2751dabaa243b0708ff/README.md#extending-the-repository)

You won't benefit from all the magic behind like the cache system, but you can implement it easily yourself if you need it (you won't need it if you code is clean, but I know we do :) )

The implementation can be something like this:

```php
    public function findBy(array $params) {
        // generate the path to call
        $path = // ...
        $data = $this->restClient->get($path);
        $hydrator = $this->sdk->getModelHydrator();
        $out = [];
        foreact ($data as $item) {
            $entityName = $item['@type'] === 'Bar' ? 'Foo\Bar' : 'Foo\Baz';
            $out[] = $hydrator->hydrate($item, $entityName);  
        }
        return $out;
    }
```

**Comment by jmfeurprier** [comment]
Fixed in v0.21.0.
